### ISSUE TITLE
runtime: frequent timeouts in {build,runBuilt,run}TestProg

### ISSUE THREAD

**Issue by bcmills** [golang/go#44422]
<details>

[2021-02-19T00:04:42-1c659f2/linux-ppc64le-buildlet](https://build.golang.org/log/202b1cb8d9185291e4feb5ebee2ddd05e12635b8)
[2021-02-19T00:04:30-b6379f1/linux-ppc64le-buildlet](https://build.golang.org/log/3fba541d43df459737b964eff06ac2beba791114)
[2021-02-19T00:04:22-09e059a/linux-ppc64le-buildlet](https://build.golang.org/log/ecc9df09e3dd890fc45cec6e6dd3feed44c0da59)
[2021-02-19T00:02:06-01f05d8/linux-ppc64le-buildlet](https://build.golang.org/log/5650936938a0ea6502add1c12e1d656edcc8b962)
[2021-02-19T00:01:17-e7ee3c1/linux-ppc64le-buildlet](https://build.golang.org/log/b9425f32f72e042dcfe6a5082554b7a841d604d1)
[2021-02-16T22:18:37-8482559/linux-ppc64le-buildlet](https://build.golang.org/log/34774e3630b79b0ff571f321cb001d96472b64c8)
[2020-12-15T21:45:05-731bb54/linux-ppc64le-buildlet](https://build.golang.org/log/2ae274e0c4748a83fe2ac7421baab7f09950bb0f)
[2020-12-15T21:04:49-129bb19/linux-ppc64le-buildlet](https://build.golang.org/log/68d6579d1f9e32dfee704e1f9904fc4e5c8e53b3)
[2020-12-15T21:01:37-685a322/linux-ppc64le-buildlet](https://build.golang.org/log/f99bab7a4e1f5695df2831422e79500ae8f17812)
[2020-12-15T20:58:17-3d64678/linux-ppc64le-buildlet](https://build.golang.org/log/7caf127c582c79793d41ddcaf880f2786cc5c9bf)
[2020-12-15T20:55:01-7cdc84a/linux-ppc64le-buildlet](https://build.golang.org/log/7c727ee0d251d2638f4b4f2f1091a80c3c370fc6)
[2020-12-14T22:39:04-663cd86/linux-ppc64le-buildlet](https://build.golang.org/log/a53d3d79dddca177ca31fa548fbe2a864f63eb69)
[2020-12-14T21:09:33-d06794d/linux-ppc64le-buildlet](https://build.golang.org/log/4cb46986e4dc3822b98e5066a8ee6cc530941464)
[2020-12-14T18:06:06-828746e/linux-ppc64le-buildlet](https://build.golang.org/log/6c8e2ad43c675b3094ce01058f1732bc4c011cad)

</details>

It's not clear to me whether the test is hung or just slow. Perhaps we should set a `GO_TEST_TIMEOUT_SCALE` on this builder and see if that helps?

CC @golang/release @bradfitz @laboger @ceseo

**Comment by laboger** [comment]
I've been looking at this problem on various systems. If I run 'go tool dist test go_test:runtime' by itself the time varies and sometimes on a power8 it times out, but never on a power9 because it is faster there. But one mysterious result is that if I set GOMAXPROCS to a smaller value, like 2, 4, or 8 then the time is small enough on the power8 so that it doesn't time out. I thought these VMs only had 2 processors each, which I thought was equivalent to running GOMAXPROCS=2.

**Comment by ceseo** [comment]
I don't know if this is related, but a while back I had a bunch of timeouts in my personal OSU power9 system. I asked them to increase it from 2 to 4 cores and 4 to 8GB RAM and now I have zero timeouts there. Maybe these builders are timing out because they have too little CPU time allocated to them in the hypervisor.

@bradfitz can correct me if I'm wrong, but IIRC, the power8 builders are 2 core, 4GB RAM only.

**Comment by laboger** [comment]
From golang.org/x/build/env/linux-ppc64le/osuosl/NOTES:
* go-le-bionic-1: (20 GB RAM, 50 GB disk, 10 cores, POWER9)
  Linux go-le-bionic-1 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:08:54
UTC 2019 ppc64le ppc64le ppc64le GNU/Linux
* go-le-bionic-2: (20 GB RAM, 50 GB disk, 10 cores, POWER8)
  Linux go-le-bionic-2 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:08:54
UTC 2019 ppc64le ppc64le ppc64le GNU/Linux

**Comment by laboger** [comment]
I>t's not clear to me whether the test is hung or just slow. Perhaps we should set a GO_TEST_TIMEOUT_SCALE on this builder and see if that helps?

@dmitshur I think we need to increase the GO_TEST_TIMEOUT_SCALE for all the linux ppc64* builders.  I can make the change but don't know how those are tested.

**Comment by dmitshur** [comment]
@laboger Sounds good. GO_TEST_TIMEOUT_SCALE is a part of builder configuration, it doesn't need tests. See [CL 158537](https://golang.org/cl/158537) for a past example.

**Comment by bcmills** [comment]
This turns out not to be specific to the ppc64le builder. I think the `runtime` tests are just getting too slow in short mode in general.

(CC @jeremyfaller)

[2021-04-28T00:49:58-c9f4350/illumos-amd64](https://build.golang.org/log/fa05413712fd23e60446b37c1684003fae5c174f)
[2021-04-27T15:50:45-88a8a27/freebsd-amd64-12_2](https://build.golang.org/log/f4c7bd60e551e98c3e8f7c834ded77ef373a895f)
[2021-04-22T04:07:38-14a18b7/illumos-amd64](https://build.golang.org/log/0cd70d3a8511cc8dc3cab8c5d1cc5409b5adfd91)
[2021-04-22T03:03:48-617a83e/netbsd-arm64-bsiegert](https://build.golang.org/log/e553c68e5f187b69684785a50006580dad6cfac3)
[2021-04-21T21:14:04-54af9fd/illumos-amd64](https://build.golang.org/log/9432b7eca607d3abd4e4af97c8ab08cdb3d5204d)
[2021-04-20T22:47:50-48e3d92/solaris-amd64-oraclerel](https://build.golang.org/log/da449072969798502a1bbe3518f7e5696336730d)
[2021-04-20T20:14:14-7473a6a/illumos-amd64](https://build.golang.org/log/74afd6e95131bca814b4edad93e094130caea0e5)
[2021-04-20T14:30:03-3ff6ff7/illumos-amd64](https://build.golang.org/log/505660a0b6c61a84eae78357a4eb9eaa1169d1f7)
[2021-04-15T12:38:11-dba2eab/illumos-amd64](https://build.golang.org/log/a50b4a1031bffefdc7cb4de5b48442c7af5faf95)
[2021-04-13T14:13:57-4b00eb7/illumos-amd64](https://build.golang.org/log/5ee97d9f919ca8afa8a771edd2949fe153c44853)
[2021-04-12T21:59:59-263e13d/illumos-amd64](https://build.golang.org/log/05e4d87cf79c1eeddc5086d9728a281c6f393697)
[2021-04-08T02:17:23-8752454/illumos-amd64](https://build.golang.org/log/68be15cb302820f3b7457a0e04a1dd4fa5c1cae9)
[2021-04-03T18:16:29-6986c02/linux-arm-scaleway](https://build.golang.org/log/1f37d244fb197e3ca21590e4194a432e834c565c)
[2021-04-02T20:26:12-35a8bbc/illumos-amd64](https://build.golang.org/log/a351a4053116d2654b36d4580b727c2042f69312)
[2021-04-02T19:27:06-34b87b4/illumos-amd64](https://build.golang.org/log/4efc86be07392d1122d1759ab4d8d673bcb67066)
[2021-04-02T16:53:18-28c5fed/illumos-amd64](https://build.golang.org/log/03e6bacaa03bf7862827d803708aa6618032d0d9)
[2021-03-25T04:14:58-ddcdbb4/illumos-amd64](https://build.golang.org/log/b2bf0e52d0c8ed56cc4eaf0c3e96f8a319b7e0dc)
[2021-03-18T21:10:31-79d03ad/linux-ppc64le-buildlet](https://build.golang.org/log/b3be1c79dffefa932f5f09d352cd19a05b3bf4de)
[2021-03-16T20:46:00-72d98df/linux-ppc64le-buildlet](https://build.golang.org/log/9a3ca80b0bc31fda4c3f49caf073a7fa8933c4b1)
[2021-03-13T18:03:52-8336c31/illumos-amd64](https://build.golang.org/log/506659aa8c086ac2f603fde568a72c4b8b8a8696)
[2021-03-11T16:37:38-0fc370c/illumos-amd64](https://build.golang.org/log/a8e6e20056ff1896eb3720587f190cf124fbbe23)
[2021-03-11T08:19:31-79e3ee5/illumos-amd64](https://build.golang.org/log/e0549fbde39e2144d3c4f39145226828030d83cd)
[2021-03-09T22:00:50-142a765/illumos-amd64](https://build.golang.org/log/a80eddeddc8d81a4cdf8b8673037a41f4787bacc)
[2021-03-07T23:02:35-b0df927/illumos-amd64](https://build.golang.org/log/e94e5146433a8b8e6de3c617054d4ccdfbc0ead2)
[2021-03-04T22:53:05-56d52e6/linux-ppc64le-buildlet](https://build.golang.org/log/5a45d361f9e05662ca0190f5f518d019c59efacd)
[2021-02-25T23:19:16-a61524d/linux-ppc64le-buildlet](https://build.golang.org/log/c63a293a545124dee377bf8127c5a19071e3d22c)
[2021-02-25T21:52:49-9a7fe19/linux-ppc64le-buildlet](https://build.golang.org/log/3d98d648190f3470849f43c910472d03792c20e1)

**Comment by bcmills** [comment]
This may be the same underlying issue as #45885.

**Comment by bcmills** [comment]
`greplogs --dashboard -md -l -e '(?ms)test timed out.*runtime_test\.(?:run|runBuil[dt]|build)TestProg.*FAIL\s+runtime' --since=2021-05-01`

<details>

[2021-11-08T16:15:01-759eaa2/solaris-amd64-oraclerel](https://build.golang.org/log/37cc1c1d88be7ef5f1c2a6ac4e1f01b55c3a2d24)
[2021-11-07T04:56:11-85493d5/solaris-amd64-oraclerel](https://build.golang.org/log/4db502a0d83b2b825af0f4a39d92d07e730480de)
[2021-11-05T22:00:37-f00b43f/solaris-amd64-oraclerel](https://build.golang.org/log/087a30b572612df1c4c802d66d1c8eb6c59f8a30)
[2021-11-05T21:28:34-90462df/solaris-amd64-oraclerel](https://build.golang.org/log/9de2a1f5b15814221a06f9414ff01ea85eb4ff85)
[2021-10-29T19:51:20-994049a/netbsd-arm64-bsiegert](https://build.golang.org/log/5de112b0f4c42b931b5ee519807789bf69756451)
[2021-10-28T03:35:34-b8f928b/netbsd-arm64-bsiegert](https://build.golang.org/log/99421b638857a6a588b54a969509a021889be57a)
[2021-10-28T03:35:04-5a3a9d8/netbsd-arm64-bsiegert](https://build.golang.org/log/711748338f911b92e2bf792f092f505646aeb1d3)
[2021-10-25T20:39:17-c26a32a/openbsd-386-68](https://build.golang.org/log/3a5d08b0caf9a04897bf775495b4d1620e3e033d)
[2021-10-16T16:27:40-640a49b/solaris-amd64-oraclerel](https://build.golang.org/log/8d1c80187fd03f8f2c974cd3ee23315d720685c8)
[2021-10-06T22:42:28-4002616/netbsd-arm64-bsiegert](https://build.golang.org/log/fe70180d76ef18a4c3c7df92347ef26bef553648)
[2021-10-02T13:52:59-07a7c6f/solaris-amd64-oraclerel](https://build.golang.org/log/4de8af0784610e7c1669c612d6ee4e64713a9c27)
[2021-10-01T23:51:16-c129af9/netbsd-arm64-bsiegert](https://build.golang.org/log/ea74cc468da91a2705d573ddb332f0731435f8c1)
[2021-10-01T17:40:49-243d65c/solaris-amd64-oraclerel](https://build.golang.org/log/8b8fdd08d49a53085288515b652f08033c4e865f)
[2021-10-01T17:18:34-dab16c1/netbsd-arm64-bsiegert](https://build.golang.org/log/8ea5e1ed120279c5b417560ae753c70cb5ce7656)
[2021-09-30T18:10:18-205640e/netbsd-arm64-bsiegert](https://build.golang.org/log/cc2dbbc169e5cb6a991eceaa1b8a4f2b7dcf1aa0)
[2021-09-28T23:59:17-435718e/netbsd-arm64-bsiegert](https://build.golang.org/log/4b3823360eac35919641d20a309452b3741f4762)
[2021-09-24T17:23:19-d5d3f80/netbsd-arm64-bsiegert](https://build.golang.org/log/cb1a57ed2c189bbcd12e5e6faca409215b6f5869)
[2021-09-24T14:52:47-217507e/netbsd-arm64-bsiegert](https://build.golang.org/log/32f834152e4afa865af9ce41d48468647532206d)
[2021-09-17T22:19:21-3fa35b5/linux-ppc64le-power9osu](https://build.golang.org/log/285fecaa70fcab3972922a246e6d5b543a9fc635)
[2021-09-13T20:17:24-a0c409c/solaris-amd64-oraclerel](https://build.golang.org/log/0d4a693274918155438bbda7d603a62712f757b4)
[2021-09-12T01:06:53-0d8a4bf/linux-riscv64-jsing](https://build.golang.org/log/bde765090d91e04aeeb2e999c1bc76ec9c3333cb)
[2021-09-10T17:11:39-5a4b9f9/solaris-amd64-oraclerel](https://build.golang.org/log/d7cb602a30b2bc2e3e2d8eab5c9af2cd242c2687)
[2021-09-10T16:51:52-5a94a90/openbsd-386-64](https://build.golang.org/log/eadc582cc35d73493da6689dffcda0232b4dd626)
[2021-09-10T14:03:35-96ab854/solaris-amd64-oraclerel](https://build.golang.org/log/7bb8b1c34ac782a4bbe41c460a5f08cc72c1c31e)
[2021-09-10T02:44:36-2091bd3/netbsd-arm64-bsiegert](https://build.golang.org/log/cd73f26399af6c5388f7b8413de1e409444de709)
[2021-09-09T04:47:41-c84f3a4/linux-riscv64-jsing](https://build.golang.org/log/3cf598493f1abdbb5228291a3438425d1c530f2b)
[2021-09-06T19:10:15-a193843/netbsd-arm64-bsiegert](https://build.golang.org/log/9831da7b3f4d60bd9b49eee550e643d80297d6e4)
[2021-09-06T03:31:43-43b0517/netbsd-arm64-bsiegert](https://build.golang.org/log/d622d143afb05e14842d274d469a002730e0a7b0)
[2021-09-01T06:18:05-5a687ee/netbsd-arm64-bsiegert](https://build.golang.org/log/39de80cfe4c57344d2d17ac9555347654e628b62)
[2021-08-28T06:22:11-044550a/solaris-amd64-oraclerel](https://build.golang.org/log/f6d03cae43cac1ce8d039f5d3ae1c8c0ddc47f57)
[2021-08-26T20:39:50-03db2c2/netbsd-arm64-bsiegert](https://build.golang.org/log/354e0cf6dfa95e36aeb9d643fc1b7e43db7fcf4b)
[2021-08-25T15:30:19-099b819/solaris-amd64-oraclerel](https://build.golang.org/log/b15e0b65190e7fac68c74ff3a0ce172d4781e08d)
[2021-08-21T11:22:07-bcd146d/solaris-amd64-oraclerel](https://build.golang.org/log/f3f5e0ea46b85bdb3f5bb20f493e0949755f1c60)
[2021-08-17T04:37:32-a304273/illumos-amd64](https://build.golang.org/log/dd29ec9bfa202625f446cf222481013243f55ba8)
[2021-08-16T22:27:47-213e157/illumos-amd64](https://build.golang.org/log/d0cad64438e4319cb87705886531b0131d0f9635)
[2021-08-16T16:58:24-a0adf91/illumos-amd64](https://build.golang.org/log/9892c2740dc5be7805392b507619eb5dcab18d86)
[2021-08-16T12:42:58-94002f6/illumos-amd64](https://build.golang.org/log/d4da7e0374e3dd1887a1f197e7f8a119a0620822)
[2021-08-16T10:46:11-3d679c6/netbsd-arm-bsiegert](https://build.golang.org/log/a15c42c36aac29a7cea2f75ecd950e6e5c897858)
[2021-08-16T08:37:46-fcdc3c0/illumos-amd64](https://build.golang.org/log/e750120d021d85411451ec164359b9fefada48ca)
[2021-08-14T15:14:57-456759b/illumos-amd64](https://build.golang.org/log/d916f5da7796889f7d161d2f2256e4b9b41c3383)
[2021-08-13T20:51:21-89a4f99/illumos-amd64](https://build.golang.org/log/5e3edb365594b4014ca46d182526c6848c2fe9ac)
[2021-08-13T20:49:48-bad1fc1/illumos-amd64](https://build.golang.org/log/293ee853b71c5719b51d37c641dfca22ea549e0b)
[2021-08-13T20:45:17-98f3d7f/illumos-amd64](https://build.golang.org/log/fd40c241814140e61909f1f20fe3dfbe5fee3807)
[2021-08-13T00:12:29-1fffedd/illumos-amd64](https://build.golang.org/log/b448c2fb39045efcb662b6561603fac4d544491c)
[2021-07-26T14:24:27-1868f82/illumos-amd64](https://build.golang.org/log/e1f733a4bfec8e31af1a11276daf47dbee1c0684)
[2021-07-26T14:24:27-1868f82/solaris-amd64-oraclerel](https://build.golang.org/log/3dd73df71b7f19804a3427e9299187353f05ea2d)
[2021-07-25T17:16:20-849b791/illumos-amd64](https://build.golang.org/log/1361b80322ffb900247d09b4c8ff61ca61566171)
[2021-07-25T17:16:20-849b791/solaris-amd64-oraclerel](https://build.golang.org/log/ccd500e0bfbbec9a2791ff6a07ee2f8bf638963c)
[2021-07-23T18:10:55-0914646/aix-ppc64](https://build.golang.org/log/6ce2197c238484318605e842178e26ea0aa1edf8)
[2021-07-23T18:10:55-0914646/illumos-amd64](https://build.golang.org/log/e494183caba66b58d19eb9eeb9acb41c168b7e7e)
[2021-07-23T18:10:55-0914646/solaris-amd64-oraclerel](https://build.golang.org/log/36b887059fdbb8556dc8febabd7124df0f5a5876)
[2021-07-22T20:47:59-052da57/freebsd-arm64-dmgk](https://build.golang.org/log/6113ddf6b41a7ab95ea1860e398953baa7285d84)
[2021-07-22T19:26:40-798ec73/solaris-amd64-oraclerel](https://build.golang.org/log/9df745b817f9710ec52b02c76a13e8dab2e3d603)
[2021-07-20T19:54:36-d568e6e/illumos-amd64](https://build.golang.org/log/793529c554c6576770e69edacf6e60ee2b9c66ec)
[2021-07-09T18:00:16-ab4085c/illumos-amd64](https://build.golang.org/log/56f2a0dcd491aaa241bcc142f4ee961e0a590e0e)
[2021-07-08T09:09:53-2ca44fe/illumos-amd64](https://build.golang.org/log/c3b0c5bbba91905af82208577884e9a57a52dc0b)
[2021-07-07T22:27:48-11f5df2/illumos-amd64](https://build.golang.org/log/1d5db1d5038e7df0f41f6bf05135b84c53c0b726)
[2021-06-29T21:54:24-f9d5095/illumos-amd64](https://build.golang.org/log/8bc246358c02433841df4512edccbbb2692e0ab4)
[2021-06-22T16:59:10-0ebd5a8/illumos-amd64](https://build.golang.org/log/c838b2139f2dafe1c268c9eda8aa9d8d2a9c8a4a)
[2021-06-15T18:42:11-219fe9d/illumos-amd64](https://build.golang.org/log/ab2ffa674536e3a470dd778c29cb3601f8f943eb)
[2021-06-04T14:31:24-c6b6211/illumos-amd64](https://build.golang.org/log/a2dc6aa89de480557ee2209f65c580a215f3f3c5)
[2021-06-01T21:14:02-567ee86/illumos-amd64](https://build.golang.org/log/a6c20346aacb39e1029f4e85c5471261eb945a9e)
[2021-05-30T02:37:38-1607c28/illumos-amd64](https://build.golang.org/log/131a0bfb01587d3579e218bcfe3a4c797a23b4d8)
[2021-05-26T16:11:00-bfd7798/illumos-amd64](https://build.golang.org/log/ff176c15936b40932b6b8b012d4402ba769aa111)
[2021-05-26T05:45:56-d050238/illumos-amd64](https://build.golang.org/log/08375ba3c04e5e119abdc99cc42f97a719268d2f)
[2021-05-14T16:42:01-3d324f1/illumos-amd64](https://build.golang.org/log/757a2cc441d14b61b7ae1350651456b5fed94c01)
[2021-05-12T17:32:39-f93b951/illumos-amd64](https://build.golang.org/log/7eb3377c456b58c1922b47406ad49bda0a0fed42)
[2021-05-12T15:23:09-0388670/illumos-amd64](https://build.golang.org/log/aec302e50989ebea2a63648b9c6706136f32efb1)
[2021-05-12T15:04:42-af0f8c1/illumos-amd64](https://build.golang.org/log/6ee8beb15df15f8dc359f2ab0ed855d97f969cb3)
[2021-05-09T17:07:01-bedf2c4/illumos-amd64](https://build.golang.org/log/b4eb2294780a0773d4de5eaa3a30fab2ca812a3f)
[2021-05-08T14:59:49-68327e1/illumos-amd64](https://build.golang.org/log/0895b00ebaaa76e446882a6ce3475e686d256f73)
[2021-05-08T05:57:31-4c8f48e/illumos-amd64](https://build.golang.org/log/9900ac3cad9ab11cfe23f62e3ea8d48fad220808)
[2021-05-07T02:17:32-d2b0311/illumos-amd64](https://build.golang.org/log/cbd3fbb833759634043056dcb91ac2a9968b4dd7)
[2021-05-06T15:33:43-54e20b5/illumos-amd64](https://build.golang.org/log/bcc99f258386592065daa94fec17c16700f8e0a5)
[2021-05-05T01:28:34-4df662f/illumos-amd64](https://build.golang.org/log/f89c7975dbc819b9b6e4de3c393fecdc021bb298)
[2021-05-04T15:03:06-86e42c2/linux-ppc64le-buildlet](https://build.golang.org/log/cebdeef298eb016344c90cdc7f5aa9eb7bbc7d98)
[2021-05-03T01:35:44-2c9f5a1/illumos-amd64](https://build.golang.org/log/3ebcc2591c1ef39ede63b250f1d535879f83a729)

</details>

**Comment by bcmills** [comment]
This is affecting at least five different builders, with many failures per month.

I think that makes it a release-blocker via #11811. All of the builders need to be passing tests reliably — if that means skipping more tests on non-longtest builders, we may need to prioritize which tests to run, or perhaps move some of the slower ones out to some package that won't also be run by users as part of `go test all` in their own modules. (Compare #48770.)

**Comment by bcmills** [comment]
Looking at the `solaris-amd64-oraclerel` timeouts, many of the process wait times are quite long (1–2 minutes).

That makes me wonder whether these timeouts are actually deadlocks — perhaps related to the ones observed in #48789?

**Comment by bcmills** [comment]
#49614 also shows an apparent deadlock with processes blocked in `runtime.syscall_wait4`. (It may be a duplicate of this issue.)

**Comment by ianlancetaylor** [comment]
It's interesting to note that in all the recent Solaris failures the runtime test times out while building a test with build tags, either from `TestCgoExternalThreadSignal` (building `testprogcgo` with `-tags=threadprof`) or `TestCheckPtr` (building `testprog` with `-gcflags=all=-d=checkptr=1`).  The `buildTestProg` function serializes all builds using a mutex.  It does a separate build for each binary (`testprog`, `testprogcgo`, ....) and the additional arguments being passed to `go build`, typically none but sometimes, as in these cases, present.  Building `testprogcgo` naturally requires invoking the C compiler a few times, and also the C linker.  So overall these extra builds could definitely slow down the test.  Still hard to see how it takes as long as it does, though.

**Comment by bcmills** [comment]
> in all the recent Solaris failures the runtime test times out while building a test with build flags

That is a strong pattern but not universal. The stack traces for the failures listed in #49614 seem to be at a point where the test is running the compiled program (and has been for over a minute), not waiting on a build process.

**Comment by ianlancetaylor** [comment]
After CL 364654 the time required to run the runtime tests on a Solaris system I have access to drops from 70 seconds to 51 seconds.  CL 364755, if submitted, will drop it faster.  I don't know if timeouts are the problem here, but they may be part of it.  But there may also be a different problem.

**Comment by bcmills** [comment]
Seems like a good start, at least. We can watch the builders for a while and see whether it is empirically fixed.

**Comment by bcmills** [comment]
The last timeout in these functions was before the aforementioned changes, so I'm provisionally marking this as fixed.
(The reduction in these timeouts did draw my attention to a _different_ `runtime` timeout, which I have filed as #49680.)

`greplogs --dashboard -md -l -e '(?ms)test timed out.*runtime_test\.(?:run|runBuilt|build)TestProg.*FAIL\s+runtime' --since=2021-11-17`

[2021-11-17T18:14:41-0981724/openbsd-386-68](https://build.golang.org/log/3345b32c0fc4f51a697b94fdd4bd24f5e0649d3e)

**Comment by bcmills** [comment]
Nope, still timing out occasionally:

`greplogs --dashboard -md -l -e '(?ms)test timed out.*runtime_test\.(?:run|runBuilt|build)TestProg.*FAIL\s+runtime' --since=2021-11-20`

[2021-11-22T20:34:40-f13fcd9/solaris-amd64-oraclerel](https://build.golang.org/log/d3bd7e0d325d2effdbc641b828b3dcf7367b374a)

**Comment by ianlancetaylor** [comment]
In 2021-11-22 timeout the only tests still running are two waiting for `go build -tags=threadprof testprogcgo` to complete.

**Comment by ianlancetaylor** [comment]
On a Solaris system I have access to, clearing the cache and running `go test -test.short -test.v -test.count=1 runtime` takes 38.5 seconds.  The last two tests to complete are `TestCgoExternalThreadSIGPROF` and `TestCgoExternalThreadSignal` which both take about 13 seconds.  One test is building the test program, and the other is waiting for that build to complete (they both build the same test program).  Actually running the test program takes a fraction of a second.

This is consistent with the builder failure just being slow and running out of time, especially since cmd/dist will be running other tests in parallel.

**Comment by bcmills** [comment]
`greplogs --dashboard -md -l -e '(?ms)test timed out.*runtime_test\.(?:run|runBuilt|build)TestProg.*FAIL\s+runtime' --since=2021-11-23`

[2021-12-02T22:06:27-06dbf61/solaris-amd64-oraclerel](https://build.golang.org/log/68f42bacc4ff9384f95d43d0093b881a7bd943c3)
[2021-12-02T20:45:47-3c6295d/openbsd-386-68](https://build.golang.org/log/d0b4675087c2263dbe12d2bb14c407fd451a02f9)
[2021-11-29T19:24:34-68da368/solaris-amd64-oraclerel](https://build.golang.org/log/1a0deea7879a51795e91a755a8eb3abdd2170a61)
[2021-11-24T15:57:53-14f2b2a/openbsd-386-68](https://build.golang.org/log/7359246850acc41dcd3fb8a3678f07f0c440da77)
[2021-11-24T15:55:56-465b402/openbsd-386-68](https://build.golang.org/log/14f5c6f69d56857f49676fcd62946bfa1dd3061b)
[2021-11-24T15:55:56-465b402/solaris-amd64-oraclerel](https://build.golang.org/log/50ac30043443350fcd41f659743f6ba01cf2bed0)

All of these timeouts also appear to be actively compiling in `TestCgoExternalThread.*`.

Perhaps some of these slower tests should be skipped in short mode?

**Comment by aclements** [comment]
> Perhaps some of these slower tests should be skipped in short mode?

Since we don't have a long builder for these, this seems like it would lose important test coverage.

> or perhaps move some of the slower ones out to some package that won't also be run by users as part of go test all in their own modules.

This seems feasible. As you pointed out, this would give it its own timeout. It would require some nontrivial refactoring to pull out the testprog infrastructure, but we could do that.

**Comment by bcmills** [comment]
`greplogs --dashboard -md -l -e '(?ms)test timed out.*runtime_test\.(?:run|runBuilt|build)TestProg.*FAIL\s+runtime' --since=2021-12-07`

[2022-01-06T23:39:43-042548b/solaris-amd64-oraclerel](https://build.golang.org/log/474d50f9d379f7a607268933e5e5f4c5c99cbb4f) (building `TestFakeTime`)
[2022-01-06T15:00:16-da7891f/netbsd-386-9_0-n1](https://build.golang.org/log/f616f20620589d8a79decd6fbf5f74a80eff508b) (running `TestCgoExecSignalMask`)
[2022-01-03T23:45:06-a893d0f/openbsd-386-68](https://build.golang.org/log/cbb89834ae133401f875000994d0fa0162a60929) (building `TestCgoExternalThreadSIGPROF`)
[2021-12-22T23:31:40-d2ce939/solaris-amd64-oraclerel](https://build.golang.org/log/1413ee984265dada0d79a07d33c9e551b38b484e) (building `TestCgoExternalThreadSignal`)
[2021-12-16T21:02:37-ea26ce7/netbsd-arm64-bsiegert](https://build.golang.org/log/c376d9e65512c568b1c29886b4b078f6e755c0dd) (building `TestFakeTime`)
[2021-12-13T19:40:29-7bdbc73/openbsd-386-68](https://build.golang.org/log/62eb623d59cc1d97129a873ac3c8e98fdf1d451f) (building `TestCgoExternalThreadSIGPROF`)
[2021-12-12T14:36:17-6b89773/solaris-amd64-oraclerel](https://build.golang.org/log/3153235fd6006f98ec0681dfd65527e6a632c7cd) (building `TestCgoExternalThreadSIGPROF`)

**Comment by aclements** [comment]
A much simpler thing we could do is to increase the timeout for the runtime test. We already do this for the cmd/go tests [1] and I don't think that's any more artificial than factoring the cgo prog tests out into their own test.

[1] Or, at least, we try to. [The code](https://cs.opensource.google/go/go/+/master:src/cmd/dist/test.go;l=363?q=registerStdTest&ss=go%2Fgo) is weird.

**Comment by bcmills** [comment]
If increasing the timeout fixes the builder flakiness, I think it's a reasonable approach — Go users who run `go test all` should in general expect to either set a very generous timeout or scale it to their particular running environment.

That said, if the problem is _just_ the timeout length, should we raise the scale factors in [`x/build/dashboard/builders.go`](https://cs.opensource.google/go/x/build/+/master:dashboard/builders.go?q=TEST_TIMEOUT_SCALE&ss=go%2Fx%2Fbuild) instead of just for the specific test?

(It would be nice if we could eliminate the `cmd/go` special case in that code too, but that seems like a separate issue...)

**Comment by ianlancetaylor** [comment]
The timeouts are from building a variant of threadprogcgo with `-tags=threadprof`.  We don't need that variant.  I'll send a CL to remove it.  That will remove the extra build time and should keep the tests from timing out.  I hope.
